### PR TITLE
Option for Signal install location (internal/external memory)

### PR DESCRIPTION
**Pull request by ssgromov**
### First time contributor checklist

- [x] I have read [how to contribute](https://github.com/signalapp/Signal-Android/blob/master/CONTRIBUTING.md) to this project
- [x] I have signed the [Contributor License Agreement](https://whispersystems.org/cla/)

### Contributor checklist

- [x] I am following the [Code Style Guidelines](https://github.com/signalapp/Signal-Android/wiki/Code-Style-Guidelines)
- [x] My contribution is fully baked and ready to be merged as is

----------

### Description

Following

https://developer.android.com/guide/topics/data/install-location.html

as originally pointed out at
https://community.signalusers.org/t/allow-installation-on-external-storage-location-sd-card/2638

Added option to allow the system to install Signal on the external storage.

Unfortunately, this feature request was closed automatically already twice.

My Signal database has reached some >7Gb size after which it is not possible to update the OS & apps without a
 
`backup Signal to SD -> uninstall Signal -> update -> install Signal again -> restore from SD backup`

sequence, which is very annoying.

I am aware of the issue
https://github.com/signalapp/Signal-Android/wiki/Data-loss-when-using-an-SD-Card-formatted-as-internal-memory
which I guess is likely caused by the official update overwriting the install option.

### CODE REVIEW

**Comment by alan-signal** [comment]
According to the [docs](https://developer.android.com/guide/topics/data/install-location)

> When your application is installed on the external storage:

> The .apk file is saved on the external storage, but all private user data, databases, optimized .dex files, and extracted native code are saved on the internal device memory.

This suggests hardly anything will be stored on the external storage. Have you tested this PR and have evidence that contradicts that documentation?

**Comment by ssgromov** [comment]
@alan-signal 

The Android documentation is not exhaustive about what 'all private user data' etc. are.

Furthermore, there are two options, namely SD card is used as (1) external or (2) internal storage. I am talking right now only about Android 9 now, I haven't tested any other OS.

Whilst option (1) may imply (I haven't tested this) that private data is not movable, option (2) allows moving private data to the internal storage on the SD (I have tested this using [this workaround](https://www.quora.com/How-do-I-let-WhatsApp-store-data-directly-to-a-micro-SD-card)). 

There is a glitch, however: Signal must be freshly installed and moved to the SD space using the system interface before setting up the account.  Otherwise, a migration of files (of an already existing account) offered by Android completes, but Signal stops recognising media that were received before the move. 

And there is a workaround, too: I just made a backup to SD, uninstalled Signal, installed fresh, moved it to SD space, and then restored from the backup. That works fine for already a week.

To recap, 
- On Android 9 one can use the manifest option proposed here without changes to the code (via an override in developer options)
- Migration cannot be done on the already set-up account, but via backup
- Only "internal storage" SD use was tested, how this works with "external storage" SD is not known

=> I guess this request may be closed without merging.

**Comment by TheStealthReporter** [comment]
I've tested this change for several months now (using a local fork, even before this PR) on a Moto G5 first with Android 8.1 (Stock ROM) and then Android 9 (LineageOS) and an "internal storage" SD card. It worked flawlessly the entire time. I can confirm the entire data is moved, not just the .apk file.

The Moto G5 Stock ROM only allows moving data to SD card if it is mounted as "internal storage" (due to security/encryption) and if `android:installLocation="auto"` is set. Forcing a move to SD is not possible if one of these conditions is not satisfied.

Formatting an SD card as "internal storage" also means that the SD is almost never removed (since all data is encrypted, it is useless without this specific phone or reformatting it and loosing all data). The restarts of services should not be a practical concern in this case. It also seems like the actual behavior differs from the android documentation a bit (exact semantics might to have changed over the years).

I will test the restart of the services in a few days with following procedure (is this procedure meaningful?):

 - unmounting the SD
 - mounting the SD
 - not starting Signal
 - let somebody else send me a message
 - see what happens and report back

If there's anything else that has to be tested, I'm happy to do so. I'm not that familiar with android development, with the codebase or what services are used.

 As this is a deal-breaker for me because of limited real internal storage, I really hope this PR gets merged. This is such a simple fix for a huge problem.

**Comment by FedericoCeratto** [comment]
Related to #4065 #7792 #7881 #7521 #4297 #9126 #8417 #8684 #8099

**Comment by FedericoCeratto** [comment]
Can we please merge the PR? Signal eats internal storage which can be very limited, making the application unusable on old phones.

**Comment by sovamind** [comment]
I'd like to also ask that the storage thing get fixed. I've got a Samsung Galaxy S7 which has 32GB of internal memory but a 400GB SD.  It's sort of ridiculous that with Signal installed I keep running out of space and getting messages from Android that I need to free up storage immediately.

**Comment by ssgromov** [comment]
> I'd like to also ask that the storage thing get fixed. I've got a Samsung Galaxy S7 which has 32GB of internal memory but a 400GB SD. It's sort of ridiculous that with Signal installed I keep running out of space and getting messages from Android that I need to free up storage immediately.

@sovamind 
if it is possible on your phone, you can override and enable the option to install to the external space (see my post above).

**Comment by klues** [comment]
On my phone I'm using the SD-card as external storage, so the suggestion of @ssgromov doesn't help me. Signal is using so much space on my phone that I'm about to being forced to uninstall it. However I've convinced so many of my friends to use Signal, so it would be really embarrassing to be forced to stop using Signal because of this little storage issue.

Please find a solution for saving data to external storage, I don't want to stop using Signal.

**Comment by LeeBinder** [comment]
@ssgromov , can you please detail how you backed up Signal, and how you restored your backup?

**Comment by ssgromov** [comment]
> @ssgromov , can you please detail how you backed up Signal, and how you restored your backup?

detailed in my post above:
> And there is a workaround, too: I just made a backup to SD, uninstalled Signal, installed fresh, moved it to SD space, and then restored from the backup. That works fine for already a week.

Here's, however, again what I did :

- make a backup (e.g. to SD or internal memory, whatever), make sure the backup will be seen by newly installed Signal
- uninstall Signal
- enable `"Force allow apps on external"` setting in Android `developer options` (it's there since v. 9 only, sorry)
- install Signal afresh
- move Signal to SD space (using Android system dialogue)
- start Signal and restore from backup

Voila, it should work.

Please remember that I only tested it in situation when the SD card is used as **internal storage extension** (not as "portable", i.e. normal" SD card that can be plugged out and, e.g., be read in PC).

**Comment by LeeBinder** [comment]
Thank you @ssgromov for the detailed steps! The part I was missing was that Signal has a built-in backup, hidden under ⋮> Settings > Chats & media > Backups. For my logic that's a tool and should be right in the⋮menu ..

BTW, the "Force allow apps on external" setting in Android dev options is present in my Android 7.1.1.

Also complete uninstall of the app might not be necessary, just deletion of all its data, because the Android system dialogue only moves the data not the entire app.

**Comment by LeeBinder** [comment]
hey @ssgromov good news: success in Android 7.1.1 to stock FAT32 SD and without uninstalling Signal :)

- enabled "Force allow apps on external" setting in Android dev options
- ran Signal's built-in backup, hidden under ⋮> Settings > Chats & media > Backups
(as a side note I OCR'ed this 30 digit backup passphrase and copied it into the clipboard..)
- via Android App Manager, removed all Signal data, then changed the app's storage location from internal storage to SD card
- re-opened Signal, pasted the 30 digit passcode, re-verified with SMS code and PIN

After quite a bit of time (many minutes) all my messages are back, and I see Signal's data on my SD in one big encrypted file under:

.android_secure/org.thoughtcrime.securesms-2.asec

👍

**Comment by LeeBinder** [comment]
@ssgromov unfortunately I had to edit [my posting above](https://github.com/signalapp/Signal-Android/pull/9015#issuecomment-646335193).. if hope you have a moment to re-read it.

But before and apart from that, could you (and anybody else if you feel supportive) please share another little detail which I did not find in your comments:

**where on your SD card does Signal now store its "big" media data - videos, audio, images etc?** 

Can you please fire up your fav file manager, look for that folder (Signal? android/data/org.thoughtcrime.securesms?), look for media files (.mp4/ .mp3/ .jpg etc.) in any sub folder(s) and share the entire path(s) here? Thank you.

**Comment by LeeBinder** [comment]
hello @ssgromov . On your **_internal_** storage, can you please go to (system root) `/data/data/org.thoughtcrime.securesms/` and check if the folders `app_parts` and `app_captures` still exist, and if yes, if they still contain any big files? How big in size is your (system root) `/data/data/org.thoughtcrime.securesms/` now after moving your Signal to SD card?

**Comment by LeeBinder** [comment]
### Instructions to have Signal Data on regular external storage SD formatted as stock FAT32

<ins>**!! Requirement: root via [Magisk](https://www.xda-developers.com/how-to-install-magisk/) !!**</ins>

**Tested in Android 7.1.1 on a Samsung phone with a TouchWiz ROM**

1. root your Android by installing [Magisk and Magisk Manager](https://www.xda-developers.com/how-to-install-magisk/).
**NOTE**: I cannot give any instructions here how you do this for your specific phone!
**DISCLAIMER**: you do this completely at your own risk!!!!
2. Open Magisk Manager > Settings > Mount Namespace Mode: toggle to “`Global Namespace`”. That will enable you to create lasting Symlinks even on your external storage FAT32 SD which should also be readable by Signal!
3. Reboot
4. Install [FX File Explorer (free)](https://play.google.com/store/apps/details?id=nextapp.fx) and open it
(Note: other root and Symlink capable file managers might work, too, or might not)
5. In FX File Explorer’s home screen, tap onto the green button with a white # labeled “System (Root)”. Navigate to `/storage` > tap into your external SD card’s mount point (in my case that’s `0000-0000`) > create a folder (structure) of your choice for Signal’s data. To reflect the path on the internal system, I created a directory path with three new folders analog to the same structure as on internal: `data/data/org.thoughtcrime.securesms`. If you prefer things simple, you could also just create 1 folder labeled e.g. “`Signal Data`” - that’s entirely up to you.
**!! ATTENTION**: it is _**paramount**_ you do this as “System (Root)” - do NOT tap onto the entry for your SD card on FX’s home screen!!!
6. For quicker navigation, create a shortcut to this folder to FX’s home screen: ⋮> bookmark icon in the top center > label it "_Signal Data on SD_"
7. Go back to FX’s home screen (back button on your phone or Home icon in FX’s menu bar’s very left) > again tap onto the green button with a white # labeled “System (Root)” > Navigate to `/data/data/org.thoughtcrime.securesms`
8. Again for quicker navigation, create another shortcut to this folder to FX’s home screen: ⋮> bookmark icon in the top center > label it "_Signal Data on Internal_"
9. Prepare to copy the folder app_parts: long tap it > tap onto “Copy” in the top bar. You will see a little white clipboard indicator in the top right.
10. Go back to FX’s home screen > tap on the Icon you labeled “_Signal Data on SD_” which should take you to external storage `data/data/org.thoughtcrime.securesms`
11. Paste the clipboard content in there which should copy the app_parts folder into there
12. Long tap onto the app_parts folder, ⋮ > Symlink
13. In the little pop-up window, tap onto the path displayed unter “Symlink Directory” > swipe the top bar until you see the # > tap onto it > navigate to `/data/data/org.thoughtcrime.securesms` (FX’s home screen with the bookmarks can’t be accessed here) > OK (leave the Symlink Name as app_parts(1) > OK > OK
14. Go back to FX’s home screen > tap on the Icon you labeled “_Signal Data on Internal_” > you should see your newly created Symlink _app_parts(1_) > rename the folder _app_parts_ to e.g. _app_parts__ > rename the Symlink _app_parts(1)_ to _app_parts_
15. Open Signal > open a conversation with some image and tap onto it > it should open maximized
16. Navigate out of Signal. Recommended: force quit it via Android
17. Go back into FX > home screen > “_Signal Data on SD_” > rename _app_parts_ to e.g. _app_parts__
18. Re-open Signal > re-open the conversation with the image and tap onto it > it should NOT open!
19. Go back into FX > “_Signal Data on SD_” (should still be open) > rename _app_parts__ to _app_parts_
20. Go back into Signal, tap onto the same image preview > it should again open maximized!!

If so you have done everything correctly AND it’s working as supposed on your phone.

21. Go back into FX > home screen > “_Signal Data on internal_” > delete the folder _app_parts__ **CAREFUL**: do NOT remove the Symlink app_parts .. ;)

22. If you also have a folder _app_captures_ in Signal Data on internal, repeat the steps for that folder!

(23. **OPTIONAL AND NOT TESTED**: the Global Namespace settings and the Symlinks might stick even if you uninstall Magisk again which will also unroot your Android. If you prefer not be rooted you are welcome to test this and report back)

**DISCLAIMER**: that’s all I can do for you. Either this works on your phone, or it doesn’t.

**Comment by ssgromov** [comment]
Hi @LeeBinder ,

sorry for taking long to answer (I see a lot of updates in the meanwhile though). 
Nice to see you have succeeded.

P.S. Perhaps, uninstalling Signal helped me (as it is supposed to wipe everything), then the files you've asked about (e.g. `app_parts`) might have landed in the right paths after moving it and restoring. Who knows... FYI, I have a not rooted phone and I do not see these anyway.

Cheers

**Comment by LeeBinder** [comment]
@ssgromov Gotcha.

Hopefully one fine day the coders will give us an easy option to choose data storage.

**Comment by ssgromov** [comment]
> Hopefully one fine day the coders will give us an easy option to choose data storage.

Yes, @LeeBinder , and I think I am also missing an option of reducing picture sizes... we do not need to send 12Mpixel pictures all the time... that would also save a LOT of space. Gotta check the issues list... maybe someone proposed that already.

**Comment by ThemGo** [comment]
@LeeBinder
Thank you very much for sharing your approach!

This is how it works with a rooted phone with SuperSU - tested in Android 6.0.1 on Sony Z3:

1. Your phone is rooted and has SuperSU app
2. Open SuperSU app: deactivate "Settings/Security/Mount namespace separation" resp. "Einstellungen/Sicherheit/Namensbereichsabtrenung mounten"
Remaining steps are the same... I copied and symlinked directory `/data/data/org.thoughtcrime.securesms/app_parts` to `/storage/<0000-0000>/datadata/signal-app_parts`

Note to 4.: You will also have to install Root Add-on for FX File Explorer, otherwise the green button with a white # labeled “System (Root)” will be missing. Not necessary though to activate Pro version test phase.

It seems to be crucial to symlink only a sub directory, I tried with another app to symlink its whole `/data/data/xxx` directory and the app was broken after restart. 
BTW app SD Maid is very useful for me to identify more crowded directories worth to be symlinked ("Storage analyzer" resp. "Speicheranalyse"). Signal was on first place - by a wide margin!

**Comment by LeeBinder** [comment]
hey @ThemGo , thanks for the hints with the Root Add-on for FX File Explorer and SD Maid! Glad my share helped you.

**Comment by LeeBinder** [comment]
@michaelzangl : https://github.com/johanw666/Signal-Android/releases

**Comment by l3u** [comment]
Please make it possible to at least have the chat data on SD storage. I really don't mind Signal to be in the phone's memory, but the data grows to several GB in no time. This is really a problem, esp. on older phones. Along with the fact that a backup can't be placed on SD storage, too …

**Comment by Tmpod** [comment]
> @michaelzangl : https://github.com/johanw666/Signal-Android/releases
> 
> ..provided you don't mind the 666..

Can this fork be installed over an existing stock Signal installation as an update? I want to backup my Signal data so I can upgrade my OS, but my database is ~8GB and looking at [this](https://github.com/signalapp/Signal-Android/issues/10003#issuecomment-697078684) comment, I simply have no way of doing it except to chuck it onto my external SD card.  
It it were possible I would get the apk, update Signal with it and perform the backup.

**Comment by grandchild** [comment]
@Tmpod yes, it can -- *provided* you are confident with the `adb` commandline tool.

Also, looking at that fork's storage code, I am not sure it will detect SD cards on all phones, since it relies on calling `getExternalFilesDirs`, which doesn't always return the actual SD card paths, but the internal storage again. So I wouldn't be too sure that it will work.

Do you have an option to upgrade to Android10 on the existing phone? It seems that for Android 10 you can now select the backup location.

**Comment by Tmpod** [comment]
> @Tmpod it might -- provided you are confident with the adb commandline tool. Don't attempt any overwrite install unless you know what that means. You might well lose your data.
> Also, looking at that fork's storage code, I am not sure it will detect SD cards on all phones, since it relies on calling getExternalFilesDirs, which doesn't always return the actual SD card paths, but the internal storage again. So I wouldn't be too sure that it will work.

mmmm, I'm pretty sure I don't want to risk it then. But then again, I have no option for the backup, it simply would be too big... This backup is preventing me from upgrading my phone (still on Nougat).

> Do you have an option to upgrade to Android10 on the existing phone? It seems that for Android 10 you can now select the backup location.

I am using Lineage and would like to continue doing so, but my phone (Redmi 4X, codename `santoni`) doesn't seem to have Lineage 17 yet (might not even get it at all, it's not a super recent model), only 16 which is Android 9.

**Comment by klues** [comment]
Yes, dear stale bot, this is still very much relevant! And much more also for Signal people like @greyson-signal @alan-signal @moxie-signal 
One of the main tasks I'm currently doing on other peoples phones is to sort out and delete Signal media data because people are complaining "My phone says I don't have memory anymore!". And I have to explain to them why it's not possible to save Signal data on SD card (which I can't). This is really embarrassing and keeping me off from advertising Signal to other people, which I've done very much in the past.

**Comment by LeeBinder** [comment]
I 100% agree with @klues Other apps can. This is NOT a biggie at all. 1422 open issues just for the Android version. Signal devs, if you don't even know how to code simple stuff, then how do you want to proceed..

**Comment by l3u** [comment]
I'm pretty sure insults won't motivate the dev very much to work on this. This is for sure still a problem, esp. for older phones. But since Signal started to auto-compresses images, the amount of space used dropped significantly (in my case from about 3.5 GB to about 800 MB, keeping 6 months of messages), and thus, this problem leveled off a bit. Still, this should be kept on the to-do list.

**Comment by LeeBinder** [comment]
Even stating the obvious truth doesn't matter anymore at this point in time of this pitiful and unnecessary saga. Not knowing what I'm talking about? Then read [my instructions](https://github.com/signalapp/Signal-Android/pull/9015#issuecomment-647058674)..

**Comment by michaelzangl** [comment]
Keep it open

**Comment by tl072** [comment]
I would agree about keeping this issue open. Android phones are a pain for a variety of reasons and having some flexibility for this will make them less painful.

**Comment by LeeBinder** [comment]
[My instructions](https://github.com/signalapp/Signal-Android/pull/9015#issuecomment-647058674) are _not_ working anymore for me on my new Nokia 6.1 in LineageOS 18.1 (Android 11) with SD formatted as Internal Storage, aka "Adoptable Storage", above all for the sake of encryption.

For comparison, WhatsApp data is being written to the merged/ fused Adoptable Storage by default. That can be achieved by following [fredsky2's instructions](https://forum.xda-developers.com/t/xposed-module-4-0-xinternalsd.2789267/post-86626861). The key is

`pm grant com.whatsapp android.permission.WRITE_EXTERNAL_STORAGE`

easiest via Terminal from one's computer via adb shell.

With Signal, the corresponding command

`pm grant org.thoughtcrime.securesms android.permission.WRITE_EXTERNAL_STORAGE`

does _not_ have any effect because (as per error thrown),

> Package org.thoughtcrime.securesms has not requested permission android.permission.WRITE_EXTERNAL_STORAGE

So this issue _could_ be solved if Signal did ask for permission to write to external storage. The default data path can still be left as `/data/data/org.thoughtcrime.securesms`. But then at least those who want to change that path, would finally have the opportunity to do so. An option, rather than a doctrine.

Devs @alan-signal @alex-signal @greyson-signal, would you please at least consider, and share your ideas how this can in some way be moved forward with as little effort for you as possible?
Thank you as always

**Comment by h0jeZvgoxFepBQ2C** [comment]
Please provide a way to store the backups to the SD card, it's eating all my disk space on my internal storage any my phone stops working because of it!

**Comment by ulph0** [comment]
@LeeBinder Hi thanks yor your post above from 2 years ago. I tried that. Didn't work for me. In step 15 i see no images. I don't have adopable storage, i have Android 11. I am wondering that there is no way to force granting unrequested  access to external storage on a rooted phone. even when signal does backups to the external storage (for which it needs access to external storage?!)

**Comment by LeeBinder** [comment]
@ulph0 not working for me with SD formatted now as Mobile Storage, either. From what I know, that's most likely due to Android 11's [scoped storage](https://www.google.com/search?sclient=psy-ab&site=&source=hp&btnG=Search&q=android+scoped+storage&oq=&gs_l=&pbx=1). Yes, Signal CAN access SD for backup, but scoped storage does NOT allow symlinking anymore as described above because folders in Data/Data/.. are pretty much each in their respective dedicated sandbox, so to speak. And the Signal dev team is not willing to change the path, claiming security concerns, from what I understand. What to say....

**Comment by Arkoprovo1996** [comment]
It can be a security concern, but I think there's no point saying security when it interferes with the usability of the app, right? As in if people can't use the app because it takes up space on their device, then they'll just uninstall it and not be secure anyways. I'm on Android 11 (dotOS) & apps can request access to do this. Organic Maps, Google Maps, Telegram, all do this. So, yea.

**Comment by woeisme** [comment]
I have now just gone back to Whatsapp over Signal as Signal developers clearly aren't concerned with client or customer feedback on their product. The app slowly fills phone space making it unusable and what's worse there's no way to delete media en masse. 

I've actively stopped advocating for Signal now as I cannot in good faith recommend a client that fails basic feature / usability in so many ways. Plus they seem to be hemorrhaging money as per their recent blog post.

This ticket has laid dormant for at least a year.

**Comment by ulph0** [comment]
it is really sad that this still isn't possible. i want to be free to decide where my data is stored. i want to access all my messages, even with small system memory. why is this closed?

**Comment by LeeBinder** [comment]
see https://github.com/signalapp/Signal-Android/pull/9015#issuecomment-678820453. The beauty of open source.

**Comment by cyberbeat** [comment]
I also like to have that feature. Threema offers such an option. That is the only feature, why I still prefer threema.
